src/nnet3/decodable-batch-looped.h

```diff
@@ -0,0 +1,358 @@
+  // It should stop enqueue computation requests, 
+  // and handle the remain requests.
+  void Stop();
+
+  // Advance fake chunks until the NnetComputer becomes stable.
+  // The NnetComputer is unstable for the first few chunks, which 
+  // means the state of NnetComputer will be changed chunk by chunk,
+  // e.g. the matrix represents cell of LSTM, the matrix represents
+  // buffers for TDNN.
+  // When the NnetComputer becomes stable, which means all of the members
+  // of NnetComputer become constant, we can get and set the state for 
+  // any sequence in batch correctly.
+  void AdvanceChunkUntilStable(int32 batch_size, std::vector<bool> *batch_first); 
+  
+  // Advance one chunk in bacth
+  // The sequence represented by any request in batch may be different with last 
+  // chunk, so we should set state for requests before inferene, and get 
+  // state for requests after inference.
+  void AdvanceChunk(const std::vector<NnetComputeRequest*> &requests);
+
+  void Compute();
+
+  bool Continue();
+  
+private:
+  const DecodableNnetBatchLoopedInfo &info_;
+
+  std::vector<NnetComputerSnapshot> snapshots_;
+
+  // Continue computation or not.
+  // It don't needs lock for multi-thread synchronization,
+  // because the assignment is atomic.
+  bool is_working_;
+  
+  // Queue contains requests waiting for computation.
+  // It needs a mutex to help achieve multi-thread synchronization.
+  // The first chunk is different with other chunks, so needs a 
+  // queue to handle requests of first chunk only.
+  typedef std::chrono::system_clock::time_point TimePoint;
+  typedef std::pair<NnetComputeRequest*, TimePoint> QueueElement;
+  std::queue<QueueElement> queue_; 
+  std::queue<QueueElement> chunk1_queue_;
+  std::mutex mtx_;
+  std::condition_variable condition_variable_;
+
+  // The thread for computation
+  std::thread work_thread_;
+
+  // The vector indicates the matrices stored state are batch first or not.
+  std::vector<bool> batch_first_;
+};
+
+
+class DecodableNnetBatchLoopedOnline : public DecodableInterface, 
+  public NotifiableNnetBatchLooped {
+public:
+  // Constructor.  'input_feature' is for the feature that will be given
+  // as 'input' to the neural network; 'ivector_feature' is for the iVectors
+  // OnlineFeatureInterface *ivector_features);
+  DecodableNnetBatchLoopedOnline(NnetBatchLoopedComputer *computer,
+                                 const TransitionModel &trans_model,
+                                 OnlineFeatureInterface *input_features,
+                                 OnlineFeatureInterface *ivector_features);
+
+  BaseFloat LogLikelihood(int32 subsampled_frame, int32 index) override;
+
+  void GetOutputForFrame(int32 subsampled_frame, VectorBase<BaseFloat> *output);
+  
+  // note: the frame argument is on the output of the network, i.e. after any
+  // subsampling, so we call it 'subsampled_frame'.
+  virtual bool IsLastFrame(int32 subsampled_frame) const;
```

Comment:
```suggestion
 bool IsLastFrame(int32 subsampled_frame) const override;
```

Also, for other overridden virtual methods that follow.